### DISCUSSION TITLE
What is the purpose of navigation awaiting in `Page.command` method?

### DISCUSSION THREAD

**Discussion by lsndr** [rubycdp/ferrum#451] [General]
I'd like to understand why this logic is present here:

https://github.com/rubycdp/ferrum/blob/5271866198dad526ca898344459f1a5afa2efafb/lib/ferrum/page.rb#L344-L353

I see that `wait` param passed in `go_to` method and some methods related to mouse events. 

So basically it means if I click on element and navigation happens after `down` event, the [click will fail](https://github.com/rubycdp/ferrum/blob/5271866198dad526ca898344459f1a5afa2efafb/lib/ferrum/mouse.rb#L63) if navigation doesn't finish within timeout despite the click was successful and led to navigation.

Shouldn't be this "wait for navigation" logic controlled by client and not included in mouse class?

**Comment by route** [comment]
Please consider Discussion for all kind of questions, issues should not be used like that

**Comment by route** [comment]
There's a global `@event` https://ruby-concurrency.github.io/concurrent-ruby/1.1.5/Concurrent/Event.html set to coordinate threads.
Whenever you click a button, it doesn't mean that the page is reloaded for example, so we must wait a bit and see if there were events in the background that kicked the page for reloading, if there were some then we block on `@event` as it will be released when the page fully loads. It might also not happen and then `@event.iteration` wouldn't change, so we proceed and return from the method call.

**Reply by lsndr** [reply]
But why do I need to wait for reloading every time I click a button? And why presence of this reloading should anyhow affect result of executed command? If I clicked and it led to reloading, which took too much time it doesn't mean that my click was unsuccessful.

**Reply by route** [reply]
Because in Ferrum unlike other JS libs I try to make things safe for end-user. If you click on a link and page starts to load something we must wait until the page is ready to play with. if you click on the link and we return ASAP you'll start playing with the page that in the process of loading content. The wait is very little though, just to see if the were some network events we are interested in.

**Reply by route** [reply]
For your question about `click` the wait is `0.1` if during this period `@event.iteration` hasn't changed we simply return, if it changed we block until the page is ready and lock released.
